ibtk/include/ibtk/IBTKInit.h

```diff
@@ -0,0 +1,125 @@
+//
+//    * Redistributions of source code must retain the above copyright notice,
+//      this list of conditions and the following disclaimer.
+//
+//    * Redistributions in binary form must reproduce the above copyright
+//      notice, this list of conditions and the following disclaimer in the
+//      documentation and/or other materials provided with the distribution.
+//
+//    * Neither the name of The University of North Carolina nor the names of
+//      its contributors may be used to endorse or promote products derived from
+//      this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef included_IBTKInit
+#define included_IBTKInit
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include "SAMRAI_config.h"
+#include "tbox/SAMRAIManager.h"
+#include "tbox/SAMRAI_MPI.h"
+#include <IBTK_config.h>
+
+#include <petscsys.h>
+#ifdef IBTK_HAVE_LIBMESH
+#include "libmesh/libmesh.h"
+#include "libmesh/reference_counter.h"
+#endif
+
+#include "ibtk/IBTK_MPI.h"
+
+namespace IBTK
+{
+/**
+ * @brief Initialization for IBAMR programs.
+ *
+ * The singleton IBTKInit class handles the initializations for PETSc, LibMesh, and SAMRAI. This object should be
+ * created using the initialize() function at the start of the main() function. The destruction of the object correctly
+ * closes the libraries.
+ *
+ */
+
+class IBTKInit
+{
+public:
+    /**
+     * The initialize function creates and returns a static IBTKInit object which initializes all pertinent libraries.
+     */
+    static IBTKInit& initialize(int argc,
+                                char** argv,
+                                IBTK_MPI::comm communicator = MPI_COMM_WORLD,
+                                char* petsc_file = nullptr,
+                                char* petsc_help = nullptr);
+
+    /**
+     * Destructor. Closes libraries appropriately.
+     */
+    ~IBTKInit();
+    /**
+     * Set maximum number of patch data entries supported by SAMRAI.
+     */
+    static void setMaxNumberPatchDataEntries(int a)
+    {
+        SAMRAI::tbox::SAMRAIManager::setMaxNumberPatchDataEntries(a);
```

Comment:
Let's omit this and let users call `SAMRAIManager::setMaxNumberPatchDataEntries()` directly if they need to increase this number.

In the initializer, let's do something like
```
    SAMRAIManager::setMaxNumberPatchDataEntries(std::max(2048, SAMRAIManager::getMaxNumberPatchDataEntries());
```
or
```
    if (SAMRAIManager::getMaxNumberPatchDataEntries() != 256)
    {
        SAMRAIManager::setMaxNumberPatchDataEntries(2048);
    }
```
(256 is the default value in SAMRAI.)

Reply:
This won't work in SAMRAI2 because once `SAMRAIManager::getMaxNumberPatchDataEntries()` has been called, you can't set with `SAMRAIManager::setMaxNumberPatchDataEntries`. I'll just preset with 2048 in the case of SAMRAI2.

In SAMRAI3 this value automatically increases if it reaches the upper bound.